tchap_red_list/__init__.py

```diff
@@ -11,36 +11,382 @@
+            remove_renewed_users_txn,
+        )
+
+        # Make the renewed users re-join the discovery room if there's one.
+        for user in users_removed:
+            await self._maybe_change_membership_in_discovery_room(user, "join")
+
+    async def _setup_db(self) -> None:
+        """Create the table needed to store the red list data.
+
+        If the module is configured to interact with the email account validity module,
+        also check that the table exists.
+        """
+
+        def setup_db_txn(txn: LoggingTransaction) -> None:
+            sql = """
+            CREATE TABLE IF NOT EXISTS tchap_red_list(
+                user_id TEXT PRIMARY KEY,
+                because_expired BOOLEAN NOT NULL DEFAULT FALSE
+            );
+            """
+            txn.execute(sql, ())
+
+            if self._config.use_email_account_validity:
+                try:
+                    txn.execute("SELECT * FROM email_account_validity LIMIT 0", ())
+                except SynapseError:
+                    raise ConfigError(
+                        "use_email_account_validity is set but no email account validity"
+                        " database table found."
+                    )
+
+        await self._api.run_db_interaction(
+            "tchap_red_list_setup_db",
+            setup_db_txn,
+        )
+
+    async def _add_to_red_list(
+        self,
+        user_id: str,
+        because_expired: bool = False,
+    ) -> None:
+        """Add the given user to the red list.
+
+        Args:
+            user_id: the user to add to the red list.
+            because_expired: whether the user is being added as a result of a change in
+                their account validity status.
+        """
+
+        def _add_to_red_list_txn(txn: LoggingTransaction) -> None:
+            sql = """
+            INSERT INTO tchap_red_list(user_id, because_expired) VALUES (?, ?)
+            """
+            txn.execute(sql, (user_id, because_expired))
+
+            self._get_user_status.invalidate((user_id,))
+
+        await self._api.run_db_interaction(
+            "tchap_red_list_add",
+            _add_to_red_list_txn,
+        )
+
+        # If there is a room used for user discovery, make them leave it.
+        await self._maybe_change_membership_in_discovery_room(user_id, "leave")
+
+    async def _update_added_after_expiring(self, user_id: str) -> None:
+        """Update a user's entry into the red list to set the because_expired boolean to
+        false.
+
+        We need this because there can be a delay between the user renewing their account
+        (from an account validity perspective) and the module actually picking up the
+        renewal, during which the user might decide to add their profile to the red list.
+
+        In this case, we set because_expired to false for the user, which will cause the
+        job that removes unexpired accounts from the red list to ignore them.
```

Comment:
suggestion: I found it tricky to understand the purpose of this function.
Maybe including some text along the lines of "Called when a user adds their profile to the red list after renewing their account, but before the module picks up the renewal. Prevents the user from being removed from the red list once the renewal is noticed." would help?

Reply:
Ah yes I was wondering if this method was too confusing. Will look into documenting/naming it better.